File: swap-nodes-in-pairs/swap-nodes-in-pairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        '''
        Given a linked list, swaps every two adjacent nodes.
        Swaps the list nodes only, not the values.
        Returns the original list
        '''
        if not head or not head.next: #List is empty or has 1 node
            return head
        p1 = head
        phead = None
        last_p = None
        while p1 is not None and p1.next is not None: #List still has at least 2 elements to swap
            if not phead:
                phead = p1.next
            pnextnext = p1.next.next
            p1.next.next = p1
            if last_p:
                last_p.next = p1.next
            p1.next = pnextnext
            last_p = p1
            p1 = p1.next
        return phead

File: swap-nodes-in-pairs/swap-nodes-in-pairs.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution 
{
    public ListNode swapPairs(ListNode head) 
    {
        // Nothing to do be done
        if (head == null || head.next == null) return head;
        
        // Create dummy node to avoid null pointer issues with head
        ListNode dummy = new ListNode(-1, head);
        ListNode current = dummy;
        
        // Iterate the list and swap nodes in pairs if possible
        while (current.next != null && current.next.next != null)
        {
            // Assign nodes to be swapped
            ListNode p1 = current.next;
            ListNode p2 = current.next.next;
            
            // Swap
            p1.next = p2.next;
            current.next = p2;
            current.next.next = p1;
            
            // Iterate forward
            current = current.next.next;
        }
            
        return dummy.next;
    }
}